src/components/views/dialogs/DeviceVerifyDialog.js

```diff
@@ -21,58 +22,261 @@ import MatrixClientPeg from '../../../MatrixClientPeg';
+        });
+        this._verifier = MatrixClientPeg.get().beginKeyVerification(
+            verificationMethods.SAS, this.props.userId, this.props.device.deviceId,
+        );
+        this._verifier.on('show_sas', this._onVerifierShowSas);
+        this._verifier.verify().then(() => {
+            this.setState({phase: PHASE_VERIFIED});
+            this._verifier.removeListener('show_sas', this._onVerifierShowSas);
+            this._verifier = null;
+        }).catch((e) => {
+            console.log("Verification failed", e);
+            this.setState({
+                phase: PHASE_CANCELLED,
+            });
+            this._verifier = null;
+        });
+    }
+
+    _onSasMatchesClick = () => {
+        this._showSasEvent.confirm();
+        this.setState({
+            phase: PHASE_WAIT_FOR_PARTNER_TO_CONFIRM,
+        });
     }
 
-    return (
-        <QuestionDialog
-            title={_t("Verify device")}
-            description={body}
-            button={_t("I verify that the keys match")}
-            onFinished={onFinished}
-        />
-    );
+    _onVerifiedDoneClick = () => {
+        this.props.onFinished(true);
+    }
+
+    _onVerifierShowSas = (e) => {
+        this._showSasEvent = e;
+        this.setState({
+            phase: PHASE_SHOW_SAS,
+        });
+    }
+
+    _renderSasVerification() {
+        let body;
+        switch (this.state.phase) {
+            case PHASE_START:
+                body = this._renderSasVerificationPhaseStart();
+                break;
+            case PHASE_WAIT_FOR_PARTNER_TO_ACCEPT:
+                body = this._renderSasVerificationPhaseWaitAccept();
+                break;
+            case PHASE_SHOW_SAS:
+                body = this._renderSasVerificationPhaseShowSas();
+                break;
+            case PHASE_WAIT_FOR_PARTNER_TO_CONFIRM:
+                body = this._renderSasVerificationPhaseWaitForPartnerToConfirm();
+                break;
+            case PHASE_VERIFIED:
+                body = this._renderSasVerificationPhaseVerified();
+                break;
+            case PHASE_CANCELLED:
+                body = this._renderSasVerificationPhaseCancelled();
+                break;
+        }
+
+        const BaseDialog = sdk.getComponent("dialogs.BaseDialog");
+        return (
+            <BaseDialog
+                title={_t("Verify device")}
+                onFinished={this._onCancelClick}
+            >
+                {body}
+            </BaseDialog>
+        );
+    }
+
+    _renderSasVerificationPhaseStart() {
+        const AccessibleButton = sdk.getComponent('views.elements.AccessibleButton');
+        const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
+        return (
+            <div>
+                <AccessibleButton
+                    element="span" className="mx_linkButton" onClick={this._onSwitchToLegacyClick}
+                >
+                    {_t("Use Legacy Verification (for older clients)")}
```

Comment:
Yeah, it is a bit. Unfortunately we didn't think to build in negotiation from the get-go, so we're a bit stuck with older clients. In reality we are probably ok though since there are not all that many clients that implement e2e right now, so hopefully this can go away fairly soon.